File: src/lessons/welcome/cells/LangtonColorsEntity.java
package lessons.welcome.cells;

import java.awt.Color;

public class LangtonColorsEntity extends jlm.universe.bugglequest.SimpleBuggle {
	Color[] allColors = {Color.white, Color.black, Color.blue, Color.cyan, Color.green, Color.orange, Color.red, 
			Color.gray, Color.magenta, Color.darkGray, Color.pink, Color.lightGray};

	/* BEGIN TEMPLATE */
public void step(char[] rule, Color[] colors) {
		/* BEGIN SOLUTION */
	    Color current = getGroundColor(); 
		for (int i=0;i<colors.length;i++) {
			if (current.equals(colors[i])) {
				switch (rule[i]) {
				case 'L': turnLeft(); break;
				case 'R': turnRight(); break;
				default:
					System.out.println("Unknown command associated to i="+i+": "+rule[i]);
				}
				
				setBrushColor(colors[(i+1) % colors.length]);
				brushDown();
				brushUp();
				
				forward();
				
				return;
			}
		}
		/* END SOLUTION */
}
	/* END TEMPLATE */

	@Override
	public void run() { 
		int nbSteps = (Integer)getParam(0);
		char[] rule = ((char[])getParam(1));
		
		Color[] colors = new Color[rule.length];
		for (int i=0; i<rule.length; i++)
			colors[i] = allColors[i];
		
		for (int i=0;i<nbSteps;i++) {
			((lessons.welcome.cells.TurmiteWorld)world).currStep = i;
			step(rule,colors);
		}
	}
}

File: src/lessons/turmites/langtoncolors/LangtonColorsEntity.c
//RemoteBuggle

void step(char* rule, Color* colors, int length);

Color allColors[] = {white, black, blue, cyan, green, orange, red, gray, magenta, darkGray, pink, lightGray};

void run(){
	int nbSteps = getParam();
	char* rule=(char*)malloc(sizeof(char)*256);
	int length = getParamLangtonColor1(rule);

	Color* colors = (Color*)malloc(sizeof(Color)*length);
	int i;
	for (i=0; i<length; i++)
		colors[i] = allColors[i];

	for (i=0;i<nbSteps;i++) {
		stepDone();
		step(rule,colors,length);
	}
	free(colors);
}

/* BEGIN TEMPLATE */
void step(char* rule, Color* colors, int length) {
	/* BEGIN SOLUTION */
	Color current = getGroundColor();
	int i;
	for (i=0;i<length;i++) {
		if (current==colors[i]) {
			switch (rule[i]) {
			case 'L':
				left();
				break;
			case 'R':
				right();
				break;
			default:
				printf("100 Unknown command associated to i=%d: %c\n",i,rule[i]);
				fflush(stdout);
				//getGame().getLogger().log("Unknown command associated to i="+i+": "+rule[i]);
			}
			setBrushColor(colors[(i+1) % length]);
			brushDown();
			brushUp();

			forward(1);

			return;
		}
	}
	/* END SOLUTION */
}
/* END TEMPLATE */
